src/evo/providertx.cpp

```diff
@@ -0,0 +1,242 @@
+    // If any of both is set, it must be valid however
+    if (pl.addr != CService() && !CheckService(pl.addr, state)) {
+        // pass the state returned by the function above
+        return false;
+    }
+
+    if (pl.nOperatorReward > 10000) {
+        return state.DoS(10, false, REJECT_INVALID, "bad-protx-operator-reward");
+    }
+
+    if (pl.collateralOutpoint.hash.IsNull()) {
+        // collateral included in the proReg tx
+        if (pl.collateralOutpoint.n >= tx.vout.size()) {
+            return state.DoS(10, false, REJECT_INVALID, "bad-protx-collateral-index");
+        }
+        CTxDestination collateralTxDest;
+        if (!CheckCollateralOut(tx.vout[pl.collateralOutpoint.n], pl, state, collateralTxDest)) {
+            // pass the state returned by the function above
+            return false;
+        }
+        // collateral is part of this ProRegTx, so we know the collateral is owned by the issuer
+        if (!pl.vchSig.empty()) {
+            return state.DoS(100, false, REJECT_INVALID, "bad-protx-sig");
+        }
+    } else if (pindexPrev != nullptr) {
+        // Referenced external collateral.
+        // This is checked only when pindexPrev is not null (thus during ConnectBlock-->CheckSpecialTx),
+        // because this is a contextual check: we need the updated utxo set, to verify that
+        // the coin exists and it is unspent.
+        Coin coin;
+        if (!GetUTXOCoin(pl.collateralOutpoint, coin)) {
+            return state.DoS(10, false, REJECT_INVALID, "bad-protx-collateral");
+        }
+        CTxDestination collateralTxDest;
+        if (!CheckCollateralOut(coin.out, pl, state, collateralTxDest)) {
+            // pass the state returned by the function above
+            return false;
+        }
+        // Extract key from collateral. This only works for P2PK and P2PKH collaterals and will fail for P2SH.
+        // Issuer of this ProRegTx must prove ownership with this key by signing the ProRegTx
+        const CKeyID* keyForPayloadSig = boost::get<CKeyID>(&collateralTxDest);
+        if (!keyForPayloadSig) {
+            return state.DoS(10, false, REJECT_INVALID, "bad-protx-collateral-pkh");
+        }
+        // collateral is not part of this ProRegTx, so we must verify ownership of the collateral
+        if (!CheckStringSig(pl, *keyForPayloadSig, state)) {
+            // pass the state returned by the function above
+            return false;
+        }
+    }
+
+    // !TODO: check for duplicate IP address or keys in the dmns manager
+
+    if (!CheckInputsHash(tx, pl, state)) {
+        return false;
+    }
+
+    return true;
+}
+
+std::string ProRegPL::MakeSignString() const
+{
+    std::ostringstream ss;
+    CTxDestination destPayout;
+    std::string strPayout;
+    if (ExtractDestination(scriptPayout, destPayout)) {
+        strPayout = EncodeDestination(destPayout);
+    } else {
+        strPayout = HexStr(scriptPayout.begin(), scriptPayout.end());
```

Comment:
Yeah this is on purpose, as the limitation is temporary (as explained in the comment at line 109). Can put an assertion here for the time being, no big difference, just one more line to change later.

Reply:
Yeah, i read it. Just looking to have the same expected behavior in both places. `MakeSignString` is part of the consensus rules. Better to align it until we implement the feature.

PD: I said throw an exception because assertions in consensus related code aren't good as could cause "kill switch" kind of problems.

Reply:
better to just remove `ExtractDestination` and sign the whole script in all cases

Reply:
added in d4be0132ca355f7db89919402871c9561d66509d